### PR TITLE
ci(kitchen): add Kitchen tests

### PR DESCRIPTION
**Pull request by n-rodriguez**
### PR progress checklist (to be filled in by reviewers)

- [ ] Changes to documentation are appropriate (or tick if not required)
- [ ] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

#### Primary type

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [x] `[ci]`       Changes to the continuous integration configuration
- [ ] `[feat]`     A new feature
- [ ] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [ ] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [ ] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type

- [ ] `[docs]`     Documentation changes
- [x] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?

No.

### Documentation checklist

- [ ] Updated the `README` (e.g. `Available states`).
- [ ] Updated `pillar.example`.

### Testing checklist

- [ ] Included in Kitchen (i.e. under `state_top`).
- [ ] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [ ] Updated the relevant test pillar.

### CODE REVIEW

**Comment by myii** [comment]
@n-rodriguez Thanks for the PR.  Unfortunately, I'm really busy at the current time but I'll get around to this PR as soon as I get a chance.  Thanks for your patience!

**Comment by ixs** [comment]
Had a quick look. Looks good to me. I'm not an expert on kitchen but hey, checks work, so that is nice. Would be a +1 from me.

THanks for the addition of Red Hat and SuSE support. Would have been good to include it in the Changelog though, because that's a nice feature worth mentioning. :-D

**Comment by myii** [comment]
@n-rodriguez I've made some progress using the `ssf-formula` and after some standardisation, this is running across all unique platforms (also using some of the latest pre-salted images):

* https://travis-ci.org/github/myii/proftpd-formula/builds/715473563

This is the main commit with proposed standardisation:

* https://github.com/myii/proftpd-formula/commit/07b9d0b288da722c3a98daf58b7d5fd173825cb0

It also includes everything to get `semantic-release` working here.  If you're OK with that, I can push the commit to this PR.

Also, are you OK if we primarily test against `master` (selecting from `1.2`-`1.7` shown above)?  There have been some discussions about future testing across this org and it makes things much easier if we are testing on `master` everywhere (although there are valid points against this as well).

**Comment by n-rodriguez** [comment]
> . If you're OK with that, I can push the commit to this PR

Let's go!

> Also, are you OK if we primarily test against `master`

Yep

> Finally, we're also using a `CODEOWNERS` file for each of our (managed) formulas. I've added you as the main code owner in the commit I've shared -- are you OK with that?

Yep

File: kitchen.yml [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+---
+# For help on this file's format, see https://kitchen.ci/
+driver:
+  name: docker
+  use_sudo: false
+  privileged: true
+  run_command: /lib/systemd/systemd
+  run_options:
+    add-host:
+      - 'localhost.localdomain:127.0.0.1'
```
Comment (daks) [inline kitchen.yml:12]: what is this option for? _(didn't find any info on kitchen doc so if you have any link, I'm interested also)_

File: proftpd/init.sls [hunk 2]
```diff
@@ -17,6 +17,7 @@ proftpd:
         - mode: 644
         - template: jinja
         - defaults:
+            modules_config: {{ salt['pillar.get']('proftpd:modules_config') }}
```
Comment (daks) [inline proftpd/init.sls:20]: those changes are affecting the formula behaviour not only kitchen so I think they should be separated in their own commit with the correct naming (feat: maybe?)

File: test/salt/pillar/debian.sls [hunk 3]
```diff
@@ -0,0 +1,71 @@
+---
```
Comment (daks) [inline test/salt/pillar/debian.sls:1]: It looks like the `debian`, `redhat` and `suse` only have minor differences between them (`IdentLookups` missing on suse, `User` and `Group` different on all), it could be interesting to see if a single pillar file would be possible. Or if it's interesting to have a `common` pillar file and only set specifics in those OS-named files.

File: kitchen.yml (hunk 1 above)
Comment (daks) [reply kitchen.yml:12]: ok, `run_options` just pass options to docker when starting the container

Comment (myii) [reply kitchen.yml:12]: @daks [Extra command-line options to pass to `docker run` when starting the container.](https://github.com/test-kitchen/kitchen-docker#run_options).

File: test/salt/pillar/debian.sls (hunk 3 above)
Comment (myii) [reply test/salt/pillar/debian.sls:1]: @daks Good catch, so after separating the common values into perhaps a `default.sls`, you'll end up with something like the following.

`debian.sls`:

```yaml
---
proftpd:
  IdentLookups: '"off"'
  User: nobody
  Group: nogroup
```

`redhat.sls`:

```yaml
---
proftpd:
  IdentLookups: '"off"'
  User: nobody
  Group: nobody
```

`suse.sls`:

```yaml
---
proftpd:
  User: ftp
  Group: ftp
```

Now if you _really_ wanted to get crazy, you could leave everything in `default.sls` and _only_ override for `suse`, something like:

`suse.sls`:

```yaml
---
proftpd:
  IdentLookups: ''
  User: ftp
  Group: ftp
```

File: proftpd/init.sls (hunk 2 above)
Comment (n-rodriguez) [reply proftpd/init.sls:20]: > those changes are affecting the formula behaviour not only kitchen so

what do you mean?

Comment (daks) [reply proftpd/init.sls:20]: I mean that in your commit there are changes to kitchen/inspec like indicated in the title `ci(kitchen): add Kitchen tests` but there also are changes which are more 'invasive' because they change the formula code not only the CI/tests. I think those changes should be in a separated commit named `feat` or `refactor`

Comment (n-rodriguez) [reply proftpd/init.sls:20]: yes... maybe. but I don't have more time to spend on this PR...

Comment (myii) [reply proftpd/init.sls:20]: @n-rodriguez No problem, I'll try to get around to this soon to finalise and merge.

**Comment by myii** [comment]
Merged, thanks @n-rodriguez!

**Comment by myii** [comment]
Checked using https://github.com/myii/ssf-formula/pull/242.

Also added commit bb78e57c383417a3a4b81848b15e7c94da560c35 to get `semantic-release` working properly.
